### PR TITLE
Issue 55685 processname optimisation

### LINKED ISSUE #55685: .ProcessName should use QueryFullProcessImageNameW or GetProcessImageFileNameW
On Windows, .ProcessName should use QueryFullProcessImageNameW or GetProcessImageFileNameW.

Presently it used NtQuerySystemInformation(SystemProcessInformation).

The problem with NtQuerySystemInformation(SystemProcessInformation) is:
 - It is tremendously slower, returning _all_ information about _all_ processes,
   instead of _just_ the name, for _one_ process. This is pretty obviously terrible, unless there are very few processes, and _perhaps_ if you only have the process id and not a handle already (two syscalls, open and query). (Too bad Windows doesn't allow query by id to cut the cost.) Enum also has an advantage, sometimes, like, if you actually want more information than just the name. But generally it seems silly.

 - It does not work with terminated processes, which NtQuerySystemInformation(SystemProcessInformation) omit.
   For example, you cannot have a generic:
    WriteLine("The process {0} ended", process.ProcessName);

  unless you call .ProcessName prior to termination to populate the CLR's cache. Such order/cache-sensitivity is also fragile but maybe ok. And that is not even possible: .NET does not expose creating processes suspended, so you will always race with a fast process that terminates quickly. So `.ProcessName` is all but unusable. We are removing our uses pretty thoroughly as a result. Though it is ok on current process, which will never be terminated.

  We have actual code like this, causing us much grief, because..complicated reasons (we actually run mostly on an alternate implementation of NT, which was enumerating terminated processes, which mostly masked these problems, but not entirely; the problem is worse on native NT).
  We are going to reduce our uses of process.ProcessName as a result.

### PR DESCRIPTION
**Pull request by SteveDunn**
Fixes https://github.com/dotnet/runtime/issues/55685.

### CODE REVIEW

File: src/libraries/System.Diagnostics.Process/src/System/Diagnostics/Process.cs
Comment (am11) [inline src/libraries/System.Diagnostics.Process/src/System/Diagnostics/Process.cs:517]: ProcessName property can be moved to Windows and Unix partials. It will avoid GetProcessName overhead on Unix, and we will only cache `_processInfo` (to load the info just once) as before.

Comment (SteveDunn) [reply src/libraries/System.Diagnostics.Process/src/System/Diagnostics/Process.cs:517]: Sorry @am11 - could you elaborate? Do you mean adding these attributes on this property?
```
        [SupportedOSPlatform("windows")]
        [SupportedOSPlatform("linux")]
```

Comment (am11) [reply src/libraries/System.Diagnostics.Process/src/System/Diagnostics/Process.cs:517]: * In `Process.Unix.cs`: original `ProcessName` as it was on main branch.
* In `Process.Windows.cs`: this modified `ProcessName` property from your PR.
* In `Process.UnknownUnix.cs`: `public string ProcessName => throw new PlatformNotSupportedException();`

This way, we won't need to introduce new method in `ProcessManaged.Unix.cs`.

File: src/libraries/System.Diagnostics.Process/src/System/Diagnostics/Process.Unix.cs
Comment (am11) [inline src/libraries/System.Diagnostics.Process/src/System/Diagnostics/Process.Unix.cs:1114]: nit: maybe inline the short description.

```c#
        /// <devdoc>
        ///    <para>Gets the friendly name of the process.</para>
        /// </devdoc>
```

File: src/libraries/System.Diagnostics.Process/src/System/Diagnostics/Process.cs
Comment (am11) [inline src/libraries/System.Diagnostics.Process/src/System/Diagnostics/Process.cs:34]: This can also be moved  to Process.Windows.cs.

File: src/libraries/System.Diagnostics.Process/src/System/Diagnostics/ProcessManager.Unix.cs
Comment (am11) [inline src/libraries/System.Diagnostics.Process/src/System/Diagnostics/ProcessManager.Unix.cs:48]: This now-unused method can be deleted. :)

File: src/libraries/System.Diagnostics.Process/src/System/Diagnostics/Process.Unix.cs
Comment (SteveDunn) [reply src/libraries/System.Diagnostics.Process/src/System/Diagnostics/Process.Unix.cs:1114]: Not sure how it ended up like that!

File: src/libraries/System.Diagnostics.Process/src/System/Diagnostics/Process.Windows.cs
Comment (am11) [inline src/libraries/System.Diagnostics.Process/src/System/Diagnostics/Process.Windows.cs:900]: We can inline caching of this field, and delete `HaveNonExitedId` and changes in `EnsureState()`. Something like:

```c#
        public string ProcessName => _processName ?? (_processName = GetProcessName())!;
```

File: src/libraries/System.Diagnostics.Process/src/System/Diagnostics/Process.cs
Comment (SteveDunn) [reply src/libraries/System.Diagnostics.Process/src/System/Diagnostics/Process.cs:34]: I tried to move it, but got compilation errors.  It's used on line 958 in this file

File: src/libraries/System.Diagnostics.Process/src/System/Diagnostics/ProcessManager.Unix.cs
Comment (SteveDunn) [reply src/libraries/System.Diagnostics.Process/src/System/Diagnostics/ProcessManager.Unix.cs:48]: I tried that, but I get a compilation error:
`8>C:\git\runtime\src\libraries\System.Diagnostics.Process\src\System\Diagnostics\Process.cs(965,51,965,65): error CS0117: 'ProcessManager' does not contain a definition for 'GetProcessName'`

File: src/libraries/System.Diagnostics.Process/src/System/Diagnostics/Process.Windows.cs
Comment (SteveDunn) [reply src/libraries/System.Diagnostics.Process/src/System/Diagnostics/Process.Windows.cs:897]: But wouldn't that lose the checks that `EnsureState` performs (like checking the state is not 'exited' and other things)?

**Comment by stephentoub** [comment]
Thanks for working on this.

Can you share performance numbers?  In particular, how does this impact the cost of a) iterating through every process and getting its name, and b) iterating through every process and getting both its name and its WorkingSet64?  I ask because this is not only changing what Win32 function is used to get the process' name, it's also changing whether retrieving the name also retrieves other state.

File: src/libraries/System.Diagnostics.Process/src/System/Diagnostics/Process.Windows.cs
Comment (am11) [reply src/libraries/System.Diagnostics.Process/src/System/Diagnostics/Process.Windows.cs:897]: It doesn't require a new state; `EnsureState(State.HaveNonExitedId)` and null checks can be inlined in this getter without modifying body of EnsureState.

File: src/libraries/System.Diagnostics.Process/src/System/Diagnostics/ProcessManager.Unix.cs
Comment (am11) [reply src/libraries/System.Diagnostics.Process/src/System/Diagnostics/ProcessManager.Unix.cs:48]: Of course three comments were related, so this alone will not work.

Comment (SteveDunn) [reply src/libraries/System.Diagnostics.Process/src/System/Diagnostics/ProcessManager.Unix.cs:48]: I took another look at this just now and I think that each `ProcessManager.*.cs` needs its own copy of this method since it is called from `EnsureState` in `Process.cs` (which I _think_ is for all platforms).
I could very well be wrong though as this is my first multi-targetted solution that I've worked on, so do feel free to jump in and make any changes!

Comment (am11) [reply src/libraries/System.Diagnostics.Process/src/System/Diagnostics/ProcessManager.Unix.cs:48]: As pointed out in other comment, if you revert the changes made in EnsureState then this method is not required on Unix.

**Comment by SteveDunn** [comment]
> Thanks for working on this.
> 
> Can you share performance numbers? In particular, how does this impact the cost of a) iterating through every process and getting its name, and b) iterating through every process and getting both its name and its WorkingSet64? I ask because this is not only changing what Win32 function is used to get the process' name, it's also changing whether retrieving the name also retrieves other state.

Very welcome! Good idea re. benchmarking.  I'm happy to do this, although I think I would need some assistance.  I'm somewhat familar with benchmarking, but [I've been having difficulties consuming these changes from a standalone app outside the repo](https://discord.com/channels/732297728826277939/732297728826277942/889985214506491904).
I'd be happy to be given instructions/guidance on how to do this, alternatively, if someone wants to jump in and do it, then I'm happy with that too (especially since I am away for a week now)

File: src/libraries/System.Diagnostics.Process/src/System/Diagnostics/Process.Windows.cs
Comment (SteveDunn) [reply src/libraries/System.Diagnostics.Process/src/System/Diagnostics/Process.Windows.cs:897]: Thank you @am11 - sorry for being slow to understand :)  Initially, I was following the pattern use for 'ProcessInfo'.  I'll do these changes now.  Would you be able to help with profiling these changes?

Comment (am11) [reply src/libraries/System.Diagnostics.Process/src/System/Diagnostics/Process.Windows.cs:897]: For benchmarks, first clone a second copy of repo. Quick way is:

```pwsh
# this is powershell

$ mv runtime runtime-pr
$ git clone runtime-pr runtime-main
$ cd runtime-main
$ git remote add dotnet https://github.com/dotnet/runtime
$ git fetch
$ git reset --hard dotnet/main
```
Then benchmark dotnet:
```powershell
$ dotnet new --install BenchmarkDotNet.Templates::0.13.1
$ dotnet new benchmark --console-app -n ProcessBenchmarks
$ cd ProcessBenchmarks

# modify Benchmarks.cs

$ dotnet run -c Release -- --job Medium --filter '*' --corerun \
    ../runtime_main/artifacts/bin/testhost/net7.0-Windows-Release-x64/shared/Microsoft.NETCore.App/7.0.0/corerun \
    ../runtime_pr/artifacts/bin/testhost/net7.0-Windows-Release-x64/shared/Microsoft.NETCore.App/7.0.0/corerun
```
( or `--job Short`, if you want to get the results real quick )

Comment (SteveDunn) [reply src/libraries/System.Diagnostics.Process/src/System/Diagnostics/Process.Windows.cs:897]: > It doesn't require a new state; `EnsureState(State.HaveNonExitedId)` and null checks can be inlined in this getter without modifying body of EnsureState.

I implemented those changes, and have this failing test:
``` xml
      <test name="System.Diagnostics.Tests.ProcessTests.ProcessName_GetNotStarted_ThrowsInvalidOperationException" type="System.Diagnostics.Tests.ProcessTests" method="ProcessName_GetNotStarted_ThrowsInvalidOperationException" time="0.1419104" result="Fail">
        <failure exception-type="Xunit.Sdk.ThrowsException">
          <message><![CDATA[Assert.Throws() Failure\r\nExpected: typeof(System.InvalidOperationException)\r\nActual:   (No exception was thrown)]]></message>
          <stack-trace><![CDATA[   at System.Diagnostics.Tests.ProcessTests.ProcessName_GetNotStarted_ThrowsInvalidOperationException() in C:\git\runtime\src\libraries\System.Diagnostics.Process\tests\ProcessTests.cs:line 1028]]></stack-trace>
        </failure>
      </test>
```

Comment (am11) [reply src/libraries/System.Diagnostics.Process/src/System/Diagnostics/Process.Windows.cs:897]: Perhaps, something like this will fix the test:
```c#
            get
            {
                if (_processName == null)
                {
                    EnsureState(State.HaveId);
                    EnsureState(State.HaveNonExitedId);

                    _processName = ProcessManager.GetProcessName(_processId, _machineName);

                    if (_processName == null)
                    {
                        throw new InvalidOperationException(SR.NoProcessInfo);
                    }
                }

                return _processName;
            }
```

Comment (SteveDunn) [reply src/libraries/System.Diagnostics.Process/src/System/Diagnostics/Process.Windows.cs]: Perfect! Added a new commit and all (local) tests pass.  I won't be able to look at this for a week or so, so please feel free to jump in!  Thanks for your help!

**Comment by SteveDunn** [comment]
@am11 / @stephentoub - are those failing checks a blocker?  They don't look related to the changes in this PR.

Also, are the benchmarks required before this PR can be approved?  I won't have time to do them, but can do them in a week or so.

**Comment by stephentoub** [comment]
> Also, are the benchmarks required before this PR can be approved?

Yes.  The purpose of this PR is improving performance.  That needs to be validated for this to be worthwhile churn, and it needs to validate that it doesn't regress other usage.

Thanks.

File: src/libraries/Common/src/Interop/Windows/Kernel32/Interop.GetProcessName.cs
Comment (adamsitnik) [inline src/libraries/Common/src/Interop/Windows/Kernel32/Interop.GetProcessName.cs:4]: Do we really need that? I would expect `nullable` to be enabled by default in the entire repo.

```suggestion
```

File: src/libraries/Common/src/Interop/Windows/Kernel32/Interop.GetProcessName.cs
Comment (adamsitnik) [inline src/libraries/Common/src/Interop/Windows/Kernel32/Interop.GetProcessName.cs:31]: It would be safer to use `SafeProcessHandle` and let the marshaller do it's job rather than using `DangerousGetHandle()`:

https://docs.microsoft.com/en-us/dotnet/api/system.runtime.interopservices.safehandle.dangerousgethandle?view=net-5.0#remarks

```suggestion
            SafeHandle hProcess,
            uint dwFlags,
            ref char lpBuffer,
            ref uint lpdwSize);

        private const int MAX_PROCESSNAME_LENGTH = 1024;

        internal static string? GetProcessName(uint processId)
        {
            using (SafeProcessHandle h = Interop.Kernel32.OpenProcess(Interop.Advapi32.ProcessOptions.PROCESS_QUERY_LIMITED_INFORMATION, false, (int)processId))
            {
                Span<char> buffer = stackalloc char[MAX_PROCESSNAME_LENGTH + 1];
                uint length = (uint)buffer.Length;

                bool queried = QueryFullProcessImageName(
                    h,
```

File: src/libraries/Common/src/Interop/Windows/Kernel32/Interop.GetProcessName.cs
Comment (adamsitnik) [inline src/libraries/Common/src/Interop/Windows/Kernel32/Interop.GetProcessName.cs:25]: the `OpenProcess` method can fail:

https://docs.microsoft.com/en-us/windows/win32/api/processthreadsapi/nf-processthreadsapi-openprocess#return-value

we should most likely handle such case

File: src/libraries/System.Diagnostics.Process/src/System/Diagnostics/Process.Unix.cs
Comment (adamsitnik) [inline src/libraries/System.Diagnostics.Process/src/System/Diagnostics/Process.Unix.cs:1112]:
```suggestion
        /// <summary>Gets the friendly name of the process.</summary>
        public string ProcessName
```

File: src/libraries/System.Diagnostics.Process/src/System/Diagnostics/Process.Windows.cs
Comment (adamsitnik) [inline src/libraries/System.Diagnostics.Process/src/System/Diagnostics/Process.Windows.cs:893]:
```suggestion
        /// <summary>Gets the friendly name of the process.</summary>
```

File: src/libraries/Common/src/Interop/Windows/Kernel32/Interop.GetProcessName.cs
Comment (adamsitnik) [inline src/libraries/Common/src/Interop/Windows/Kernel32/Interop.GetProcessName.cs:21]: where does this value come from? Could you please add a comment with a link to a doc?

File: src/libraries/System.Diagnostics.Process/src/System/Diagnostics/Process.UnknownUnix.cs
Comment (adamsitnik) [inline src/libraries/System.Diagnostics.Process/src/System/Diagnostics/Process.UnknownUnix.cs:88]: throwing `PlatformNotSupportedException` would be a change in the behavior?

```suggestion
        /// <summary>Gets the friendly name of the process.</summary>
        public string ProcessName
        {
            get
            {
                EnsureState(State.HaveProcessInfo);
                return _processInfo!.ProcessName;
            }
        }
```

File: src/libraries/System.Diagnostics.Process/src/System/Diagnostics/Process.Windows.cs
Comment (adamsitnik) [inline src/libraries/System.Diagnostics.Process/src/System/Diagnostics/Process.Windows.cs:917]: When `ProcessManager.GetProcessName` calls `Interop.Kernel32.GetProcessName` and passes the result to `NtProcessInfoHelper.GetProcessShortName(processName)`

https://github.com/dotnet/runtime/blob/38e3d142f3409c6e71dbf67860d0379940301be0/src/libraries/System.Diagnostics.Process/src/System/Diagnostics/ProcessManager.Windows.cs#L110-L111

which returns an empty string for empty spans:

https://github.com/dotnet/runtime/blob/21f9b736b7a915c80f8c4df74fac4905bb9bf2fd/src/libraries/System.Diagnostics.Process/src/System/Diagnostics/ProcessManager.Win32.cs#L422-L427

so `_processName` is never going to be `null` here?

File: src/libraries/System.Diagnostics.Process/src/System/Diagnostics/ProcessManager.Windows.cs
Comment (adamsitnik) [inline src/libraries/System.Diagnostics.Process/src/System/Diagnostics/ProcessManager.Windows.cs:111]:
```suggestion
                string? processName = Interop.Kernel32.GetProcessName((uint)processId);
                if (processName is not null)
                {
                    return NtProcessInfoHelper.GetProcessShortName(processName);
                }
```

**Review by adamsitnik — CHANGES_REQUESTED** [review]
Overall looks good, but we need to improve handling of the edge cases (please see my comments). Thank you for your contribution @SteveDunn !

File: src/libraries/System.Diagnostics.Process/src/System/Diagnostics/Process.Windows.cs
Comment (SteveDunn) [reply src/libraries/System.Diagnostics.Process/src/System/Diagnostics/Process.Windows.cs:915]: Thanks @adamsitnik . I have committed your suggested change.

File: src/libraries/Common/src/Interop/Windows/Kernel32/Interop.GetProcessName.cs
Comment (SteveDunn) [reply src/libraries/Common/src/Interop/Windows/Kernel32/Interop.GetProcessName.cs:20]: Yes, good idea. I searched around the repo and grabbed it from somewhere.  I'll search around again and make a note in the code!

Comment (SteveDunn) [reply src/libraries/Common/src/Interop/Windows/Kernel32/Interop.GetProcessName.cs:20]: @adamsitnik - I can't quite remember where I found this, but on reflection, it doesn't look adequate. `MAX_PATH` in `minwindef.h`  is set to `260`, but of course, it's possible to enable long path support, but I don't know if long path support affects actual allowable filename lengths.

It would seem odd to allocate megabytes (especially on the stack) just to allow for the longest name ever imaginable, but the alternative also seems odd and unnecessary; start with a small buffer and continually increase it if the call to `QueryFullProcessImageName` writes to `ref length` with a value of our max buffer size.

Comment (adamsitnik) [reply src/libraries/Common/src/Interop/Windows/Kernel32/Interop.GetProcessName.cs:20]: We should handle `ERROR_INSUFFICIENT_BUFFER` in similar way we do it for other path-related APIs. Since handling it is a PITA I am going to push a fix to your branch, I hope you don't mind.

**Review by adamsitnik — APPROVED** [review]
LGTM, once again thank you @SteveDunn !

**Pull request merged**
